src/util/circular_buffer.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+
+#include "redismodule.h"
+#include "circular_buffer.h"
+#include "src/rmalloc.h"
+#include "src/util/likely.h"
+
+#include <stdatomic.h>
+
+// Circular buffer structure.
+// The buffer is of fixed size.
+// Items are removed by order of insertion, similar to a queue.
+struct _CircularBuffer {
+	char *read;                   // read data from here
+	_Atomic uint64_t write;       // write offset into data
+	size_t item_size;             // item size in bytes
+	_Atomic uint64_t item_count;  // current number of items in buffer
+	uint64_t item_cap;            // max number of items held by buffer
+	char *end_marker;             // marks the end of the buffer
+	char data[];                  // data
+};
+
+// Creates a new circular buffer, with `cap` items of size `item_size`
+CircularBuffer CircularBuffer_New(size_t item_size, uint cap) {
+	CircularBuffer cb = rm_calloc(1, sizeof(_CircularBuffer) + item_size * cap);
+
+	cb->read       = cb->data;                      // initial read position
+	cb->write      = ATOMIC_VAR_INIT(0);            // write offset into data
+	cb->item_cap   = cap;                           // buffer capacity
+	cb->item_size  = item_size;                     // item size
+	cb->item_count = ATOMIC_VAR_INIT(0);            // no items in buffer
+	cb->end_marker = cb->data + (item_size * cap);  // end of data marker
+
+	return cb;
+}
+
+// Returns the number of items in the buffer
+uint64_t CircularBuffer_ItemCount(CircularBuffer cb) {
+	RedisModule_Assert(cb != NULL);
+
+	return cb->item_count;
+}
+
+// Returns buffer capacity
+uint64_t CircularBuffer_Cap(CircularBuffer cb) {
+	RedisModule_Assert(cb != NULL);
+
+	return cb->item_cap;
+}
+
+// Returns the size of each item in the buffer
+uint CircularBuffer_ItemSize(const CircularBuffer cb) {
+	return cb->item_size;
+}
+
+// Returns true if buffer is empty
+inline bool CircularBuffer_Empty(const CircularBuffer cb) {
+	RedisModule_Assert(cb != NULL);
+
+	return cb->item_count == 0;
+}
+
+// Returns true if buffer is full
+inline bool CircularBuffer_Full(const CircularBuffer cb) {
+	RedisModule_Assert(cb != NULL);
+
+	return cb->item_count == cb->item_cap;
+}
+
+// Adds an item to buffer.
+// Returns 1 on success, 0 otherwise
+// This function is thread-safe and lock-free
+int CircularBuffer_Add(CircularBuffer cb, void *item) {
+	RedisModule_Assert(cb   != NULL);
+	RedisModule_Assert(item != NULL);
+
+	// atomic update buffer item count
+	// do not add item if buffer is full
+	uint64_t item_count = atomic_fetch_add(&cb->item_count, 1);
+	if (unlikely(item_count >= cb->item_cap)) {
+		cb->item_count = cb->item_cap;
```

Comment:
This assignment is not atomic, you aren't taking into account other threads that might be adding or removing items.
You should atomic decrease one from the counter.

Reply:
I think you're right. Will make this assignment atomic.